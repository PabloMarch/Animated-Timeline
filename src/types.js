// @flow

// Animation engine instance
export type AnimationEngine = Object

// Animation attributes
export type attributes = Object

// Controls for time-based execution (start, stop, reverse, reset, restart)
export type ctrl = () => void

// Time-based execution methods
export type controller = {
  start: ctrl,
  stop: ctrl,
  restart: ctrl,
  reverse: ctrl,
  reset: ctrl,
}

// Props dispatched to lifecycle hooks
export type lifecycleHookProps = {
  completed: boolean,
  progress: number,
  duration: number,
  remaining: number,
  reversed: boolean,
  currentTime: number,
  began: boolean,
  paused: boolean,
  controller: controller,
}

// Lifecycle hook signature
export type fn = (props: lifecycleHookProps) => void

// Animation lifecycle hooks
export type lifecycle = {
  onStart: fn,
  tick: fn,
  onUpdate: fn,
  onComplete: fn,
}
