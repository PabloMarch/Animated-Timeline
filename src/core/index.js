import { bezier } from './bezier'
import { easings } from './easing'
import {
  animated,
  getValue,
  createTimeline,
  random
} from './engine'

export {
  bezier,
  easings,
  animated,
  getValue,
  createTimeline,
  random
}