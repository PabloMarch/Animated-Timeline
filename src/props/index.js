// @flow

import PropTypes from "prop-types";

import { noop } from "../utils/noop";

import type { seekCtrl } from "../types";

export const defaultCommonProps = () => ({
  // controls for time-based execution
  start: false,
  stop: false,
  reverse: false,
  reset: false,
  restart: false,

  seek: (ctrl: seekCtrl) => ctrl.default(1),

  // Animation lifecycle
  lifecycle: {
    onUpdate: noop,
    onStart: noop,
    onComplete: noop,
    callFrame: noop
  }
});

export const defaultPropTypes = () => ({
  start: PropTypes.bool,
  stop: PropTypes.bool,
  reverse: PropTypes.bool,
  reset: PropTypes.bool,
  restart: PropTypes.bool,

  seek: PropTypes.func,

  lifecycle: PropTypes.shape({
    onUpdate: PropTypes.func,
    onStart: PropTypes.func,
    onComplete: PropTypes.func,
    callFrame: PropTypes.func
  })
});
