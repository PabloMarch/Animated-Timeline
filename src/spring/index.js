const rebound = require('rebound')

function createSpring(friction, tension) {
  const springSystem = new rebound.SpringSystem()
  let spring = springSystem.createSpring(friction, tension)
  let id = null

  spring.init = (prevState, el, prop, options) => {
    // el.style.transform = prop + '(' + prevState + ')'

    spring.addListener({
      onSpringUpdate: spr => {
        var val = spr.getCurrentValue()
        // val = rebound.MathUtil.mapValueInRange(val, from[0], from[1], to[0], to[1]);
        if (prop === 'rotate') val = String(val).concat('deg')

        id = window.requestAnimationFrame(
          () => (el.style.transform = prop + '(' + val + ')')
        )
      },
    })
  }

  spring.stop = () => id && window.cancelAnimationFrame(id)

  spring.setValue = spring.setEndValue

  return spring
}

class App extends Component {
  state = {
    value: 1,
  }

  componentDidMount() {
    spring.init(this.state.value, '#one', 'scale', {
      mapValues: {
        from: [],
        to: []
      },
      interpolateColors: {
        value: 1,
        startColor: 'red',
        endColor: 'blue',
        from: [0, 200],
        rgb: false,
      }
    })
  }

  render() {
    return (
      <div
        id="one"
        onMouseUp={() => spring.setValue(4)}
        onMouseDown={() => spring.setValue(3)}
      >
        Hey, Nitin!
      </div>
    )
  }
}

// Spring system for the animatione engine

const spring = createSpring(friction, tension)

const mapValues = {
  from: [0, 1],
  to: [0, 0.5]
}

const interpolateColors = {
  value: 1,
  startColor: 'red',
  endColor: 'blue',
  from: [0, 200],
  rgb: false,
}

componentDidMount() {
  spring.init(
    this.state.value, // Initial value for the animation property (scale(1))
    element, // refs or id or classname
    properties, // 'scale'
    { mapValues, interpolateColors }
  )
}

handleUp() {
  spring.setValue(5)
}

handleDown() {
  spring.setValue(3)
}

<div id="hello-world" onMouseUp={handleUp} onMouseDown={handleDown}>
  Hello World
</div>
