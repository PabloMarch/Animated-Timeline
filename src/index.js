import { createCurve } from './utils/createCurve'
import { Keyframes } from './core/keyframes'
import { transition, times, startAfter, startBefore } from './utils/properties'
import { getEasings } from './utils/getEasings'
import { getTransforms } from './core/engine'
import { Timeline } from './core/timeline'
import { createMover } from './core/createMover'
import { Animate } from './components/Animate'

// Helpers can be shared across instances of Timeline and Playback components and are used in creating animation values. (from - to, changin color values, creating bezier curves, sequencing by a offset value)
export const helpers = {
  createCurve,
  transition,
  times,
  startAfter,
  startBefore,
  getEasings,
  getTransforms
}

export { Timeline, Keyframes, Animate, createMover }
