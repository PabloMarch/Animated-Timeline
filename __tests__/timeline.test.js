import { createTimeline } from '../src'

describe('createTimeline', () => {
  it('Should create a timeline instance', () => {
    const props = {
      delay: 200,
      direction: 'alternate',
      speed: 0.2,
      duration: 1000
    }

    const timeline = createTimeline(props)

    expect(typeof timeline).toEqual('object')
    expect(timeline).toMatchSnapshot()
    expect(timeline.delay).toBe(200)
    expect(timeline.direction).toBe('alternate')
  })
})
