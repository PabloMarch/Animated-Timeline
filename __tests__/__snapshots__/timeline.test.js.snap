// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTimeline Should create a timeline instance 1`] = `
Object {
  "a": [Function],
  "abbr": [Function],
  "address": [Function],
  "altGlyph": [Function],
  "altGlyphDef": [Function],
  "altGlyphItem": [Function],
  "animatables": Array [],
  "animate": [Function],
  "animateColor": [Function],
  "animateMotion": [Function],
  "animateTransform": [Function],
  "animation": [Function],
  "animations": Array [],
  "area": [Function],
  "article": [Function],
  "aside": [Function],
  "audio": [Function],
  "autoplay": false,
  "b": [Function],
  "base": [Function],
  "bdi": [Function],
  "bdo": [Function],
  "began": false,
  "blockquote": [Function],
  "body": [Function],
  "br": [Function],
  "button": [Function],
  "cancel": [Function],
  "canvas": [Function],
  "caption": [Function],
  "children": Array [],
  "circle": [Function],
  "cite": [Function],
  "clipPath": [Function],
  "code": [Function],
  "col": [Function],
  "colgroup": [Function],
  "color-profile": [Function],
  "completed": false,
  "currentTime": 0,
  "cursor": [Function],
  "data": [Function],
  "datalist": [Function],
  "dd": [Function],
  "defs": [Function],
  "del": [Function],
  "delay": 200,
  "desc": [Function],
  "details": [Function],
  "dfn": [Function],
  "dialog": [Function],
  "direction": "alternate",
  "discard": [Function],
  "div": [Function],
  "dl": [Function],
  "dt": [Function],
  "duration": 0,
  "ellipse": [Function],
  "em": [Function],
  "embed": [Function],
  "feBlend": [Function],
  "feColorMatrix": [Function],
  "feComponentTransfer": [Function],
  "feComposite": [Function],
  "feConvolveMatrix": [Function],
  "feDiffuseLighting": [Function],
  "feDisplacementMap": [Function],
  "feDistantLight": [Function],
  "feDropShadow": [Function],
  "feFlood": [Function],
  "feFuncA": [Function],
  "feFuncB": [Function],
  "feFuncG": [Function],
  "feFuncR": [Function],
  "feGaussianBlur": [Function],
  "feImage": [Function],
  "feMerge": [Function],
  "feMergeNode": [Function],
  "feMorphology": [Function],
  "feOffset": [Function],
  "fePointLight": [Function],
  "feSpecularLighting": [Function],
  "feSpotLight": [Function],
  "feTile": [Function],
  "feTurbulence": [Function],
  "fieldset": [Function],
  "figcaption": [Function],
  "figure": [Function],
  "filter": [Function],
  "finish": [Function],
  "font": [Function],
  "font-face": [Function],
  "font-face-format": [Function],
  "font-face-name": [Function],
  "font-face-src": [Function],
  "font-face-uri": [Function],
  "footer": [Function],
  "foreignObject": [Function],
  "form": [Function],
  "frameLoop": [Function],
  "g": [Function],
  "getAnimationProgress": [Function],
  "getAnimationProgressByElement": [Function],
  "getAnimationTime": [Function],
  "getAnimationTimeByElement": [Function],
  "getAnimations": [Function],
  "getComputedTiming": [Function],
  "getCurrentTime": [Function],
  "getCurrentTimeByElement": [Function],
  "glyph": [Function],
  "glyphRef": [Function],
  "h1": [Function],
  "h2": [Function],
  "h3": [Function],
  "h4": [Function],
  "h5": [Function],
  "h6": [Function],
  "handler": [Function],
  "hatch": [Function],
  "hatchpath": [Function],
  "head": [Function],
  "header": [Function],
  "hgroup": [Function],
  "hkern": [Function],
  "hr": [Function],
  "html": [Function],
  "i": [Function],
  "iframe": [Function],
  "image": [Function],
  "img": [Function],
  "input": [Function],
  "ins": [Function],
  "iterations": 1,
  "kbd": [Function],
  "keygen": [Function],
  "label": [Function],
  "legend": [Function],
  "li": [Function],
  "line": [Function],
  "linearGradient": [Function],
  "link": [Function],
  "listener": [Function],
  "main": [Function],
  "map": [Function],
  "mark": [Function],
  "marker": [Function],
  "mask": [Function],
  "math": [Function],
  "menu": [Function],
  "menuitem": [Function],
  "mesh": [Function],
  "meshgradient": [Function],
  "meshpatch": [Function],
  "meshrow": [Function],
  "meta": [Function],
  "metadata": [Function],
  "meter": [Function],
  "missing-glyph": [Function],
  "mpath": [Function],
  "nav": [Function],
  "noscript": [Function],
  "object": [Function],
  "offset": 0,
  "ol": [Function],
  "onComplete": [Function],
  "onStart": [Function],
  "onUpdate": [Function],
  "oncancel": [Function],
  "onfinish": Promise {},
  "optgroup": [Function],
  "option": [Function],
  "output": [Function],
  "p": [Function],
  "param": [Function],
  "path": [Function],
  "pattern": [Function],
  "paused": true,
  "picture": [Function],
  "polygon": [Function],
  "polyline": [Function],
  "pre": [Function],
  "prefetch": [Function],
  "progress": 0,
  "q": [Function],
  "radialGradient": [Function],
  "rb": [Function],
  "rect": [Function],
  "remaining": 2,
  "reset": [Function],
  "restart": [Function],
  "reverse": [Function],
  "reversed": false,
  "rp": [Function],
  "rt": [Function],
  "rtc": [Function],
  "ruby": [Function],
  "s": [Function],
  "samp": [Function],
  "script": [Function],
  "section": [Function],
  "seek": [Function],
  "select": [Function],
  "sequence": [Function],
  "set": [Function],
  "setSpeed": [Function],
  "slot": [Function],
  "small": [Function],
  "solidColor": [Function],
  "solidcolor": [Function],
  "source": [Function],
  "span": [Function],
  "speed": 1,
  "start": [Function],
  "stop": [Function],
  "strong": [Function],
  "style": [Function],
  "sub": [Function],
  "summary": [Function],
  "sup": [Function],
  "svg": [Function],
  "switch": [Function],
  "symbol": [Function],
  "table": [Function],
  "tbody": [Function],
  "tbreak": [Function],
  "td": [Function],
  "template": [Function],
  "text": [Function],
  "textArea": [Function],
  "textPath": [Function],
  "textarea": [Function],
  "tfoot": [Function],
  "th": [Function],
  "thead": [Function],
  "time": [Function],
  "title": [Function],
  "tr": [Function],
  "track": [Function],
  "tref": [Function],
  "tspan": [Function],
  "u": [Function],
  "ul": [Function],
  "unknown": [Function],
  "use": [Function],
  "var": [Function],
  "video": [Function],
  "view": [Function],
  "vkern": [Function],
  "wbr": [Function],
}
`;
